src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -247,20 +247,38 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
             resolver.resolve_str_path_error(DUMMY_SP, &path_str, ns, module_id)
         });
         debug!("{} resolved to {:?} in namespace {:?}", path_str, result, ns);
-        let result = match result {
-            Ok((_, Res::Err)) => Err(()),
-            x => x,
+        let result = match result.map(|(_, res)| res) {
+            Ok(Res::Err) | Err(()) => {
+                // resolver doesn't know about true and false so we'll have to resolve them
+                // manually as bool
+                if let Some((_, res)) = is_bool_value(path_str, ns) { Ok(res) } else { Err(()) }
+            }
+            Ok(res) => Ok(res.map_id(|_| panic!("unexpected node_id"))),
         };
 
-        if let Ok((_, res)) = result {
-            let res = res.map_id(|_| panic!("unexpected node_id"));
-            // In case this is a trait item, skip the
-            // early return and try looking for the trait.
-            let value = match res {
-                Res::Def(DefKind::AssocFn | DefKind::AssocConst, _) => true,
-                Res::Def(DefKind::AssocTy, _) => false,
+        if let Ok(res) = result {
+            match res {
+                Res::Def(DefKind::AssocFn | DefKind::AssocConst, _) => {
+                    if ns != ValueNS {
+                        return Err(ResolutionFailure::WrongNamespace(res, ns).into());
+                    }
+                    // Fall through: In case this is a trait item, skip the
+                    // early return and try looking for the trait.
+                }
+                Res::Def(DefKind::AssocTy, _) => {
+                    if ns != TypeNS {
+                        return Err(ResolutionFailure::WrongNamespace(res, ns).into());
+                    }
+                    // Fall through: In case this is a trait item, skip the
+                    // early return and try looking for the trait.
+                }
                 Res::Def(DefKind::Variant, _) => {
-                    return handle_variant(cx, res, extra_fragment);
+                    if extra_fragment.is_some() {
+                        return Err(ErrorKind::AnchorFailure(
+                            AnchorFailure::RustdocAnchorConflict(res),
+                        ));
+                    }
+                    return handle_variant(cx, res);
```

Comment:
Is this really more clear than before? You've now intermingled the implementation 'variants use an anchor' with the logic 'fall back to variants if a resolution wasn't found'.

Reply:
Maybe a better approach (not necessarily here) is for _all_ of these anchor checks to happen outside `resolve()`, and not have it take `extra_fragment` at all. Then it would be the caller's responsibility to make sure that `extra_fragment` is `None` if the returned fragment exists.

The difference between that and this is the call would only have to happen in one place in `resolve_link`, not everywhere it's called.

Reply:
I was hoping to avoid passing `extra_fragment` one function down, but yeah, you are right about splitting `handle_variant` up like this is not the best idea.